Emotes and Subtles and SubtleOOC Is Blocked by Walls

Superlagg opened the pull request:
# Description

Emotes, subtles, and subtleOOCs are now blocked by walls and doors. This means people won't be able to hear you doing it in the broom closet! Unless you speak, speaking still goes through walls.

Also adds in some debug bats that spam up the chat. Pretty much only useful if you want to test chat stuff, like I did!

Also adds in functionality to let random barks use a chat mode other than Speak. This was largely to test the other chat modes with those awful chat bats.

---

<!--
This is default collapsed, readers click to expand it and see all your media
The PR media section can get very large at times, so this is a good way to keep it clean
The title is written using HTML tags
The title must be within the <summary> tags or you won't see it
-->

<details><summary><h1>Media</h1></summary>
<p>
https://www.youtube.com/watch?v=2_h8Ouw4A3k
</p>
</details>

---

# Changelog

:cl:
- tweak: Emotes, Subtles, SubtleOOCs, and LocalerOOCs are now blocked by walls and doors.
- add: Some debug bats that sit and spam chat, for testing purposes.

Superlagg committed:
Emotes and Subtles and LocalerOOC is blocked by walls
Adds LocalerOOC, which is like LOOC, but blocked by walls
Adds several flavors of debug chatterboys which are very chatty
Adds support for random barks to use other chat modes than speech

LucasTheDrgn commented:
I like a lot of what's going on with this, but I do wonder about how this affects emotes that should be audible, and thus not respect LOS. The answer might just be "yet another channel for audible emotes" and adjusting preset emotes to use that channel. Things like knocking, screaming, laughing, death gasp, etc. I feel like we might be running out of prefixes though. Either way, good work!

Superlagg commented:
> I like a lot of what's going on with this, but I do wonder about how this affects emotes that should be audible, and thus not respect LOS. The answer might just be "yet another channel for audible emotes" and adjusting preset emotes to use that channel. Things like knocking, screaming, laughing, death gasp, etc. I feel like we might be running out of prefixes though. Either way, good work!

If the emotes play a sound, then the sound should still be audible. The text won't though, that'd probably need some more code, but definitely doable.

Forzii commented:
I'm somewhat against adding an additional OOC channel - it's increasingly painful to tab through the various channels (my preferred way to switch my default channel) and I'm not sure the use case vs LOOC/subtleOOC. If anything, maybe LOOC could just get blocked by walls? 

Also, any unique files added should probably go in Floof or Floofstation folders (we haven't been super consistent as a community). 

This is super neat! Does stuff get blocked by windows too or just walls and opaque doors?

Mnemotechnician commented:
While I agree with the emote occlusion part, I do not think we need another OOC channel. We already have subtle OOC ported from the den which serves the perpose of being localler OOC.

Mnemotechnician commented on Content.Server/Chat/Systems/ChatSystem.cs:
```diff
@@ -1010,22 +1109,33 @@ private Dictionary<ICommonSession, ICChatRecipientData> GetRecipients(EntityUid
+
+            var inRange = distance < voiceGetRange;
+
+            var isVisible = observer || _interactionSystem.InRangeUnobstructed(source, playerEntity, voiceGetRange);
```
This should use InRangeUnoccluded instead in order to check for *occlusion*. Also, you don't need to calculate distance here as InRangeUnoccluded performs a range check (as the name suggests)

Superlagg replied:
Changed to using InRangeUnoccluded! Didn't see that proc there, thanks! Also the separate distance check is to allow things that arent blocked by walls to only check for range.

Mnemotechnician replied:
Oh, yea, I see, you're right.

Tk420634 commented:
> I'm somewhat against adding an additional OOC channel - it's increasingly painful to tab through the various channels (my preferred way to switch my default channel) and I'm not sure the use case vs LOOC/subtleOOC. If anything, maybe LOOC could just get blocked by walls?
> 
> Also, any unique files added should probably go in Floof or Floofstation folders (we haven't been super consistent as a community).
> 
> This is super neat! Does stuff get blocked by windows too or just walls and opaque doors?

I brought this up some to Dan last night.  SS13 has had LOOC for some time, at this point likely the better part of 8 or 9 years.  It is blocked by walls, line of sight and all that.  I have never, in all my time, felt like I needed a SubtleLOOC or wanted people outside of my field of view to be picking up my LOOCs.  Usually loocs are done for silly little things that are relevent only in the here and now with those that can see the person.  

SubtleLOOC feels like bloat to fix the symptom that your LOOC goes through walls, and LOOC going thru walls is just bad imo, because it can drag people into your looc wiggling that may not even be able to see you.  I firmly believe it would be better to not have a SubtleLOOC, or an LOOC that goes thru walls at all.  Dial it back to JUST traditional LOOC.  

Less thought on needing to tab thru chat functions (which is a whole different basket of eggs, I think THAT thing needs a pop up to toggle what channels you can even tab thru to begin with, but that's another conversation entirely) more on raw simplicity.  LOOC is used here in exactly the same way it is in 13, no reason for it to handle in some weird complex multi-channel way.

Forzii commented:
> > I'm somewhat against adding an additional OOC channel - it's increasingly painful to tab through the various channels (my preferred way to switch my default channel) and I'm not sure the use case vs LOOC/subtleOOC. If anything, maybe LOOC could just get blocked by walls?
> > Also, any unique files added should probably go in Floof or Floofstation folders (we haven't been super consistent as a community).
> > This is super neat! Does stuff get blocked by windows too or just walls and opaque doors?
> 
> I brought this up some to Dan last night. SS13 has had LOOC for some time, at this point likely the better part of 8 or 9 years. It is blocked by walls, line of sight and all that. I have never, in all my time, felt like I needed a SubtleLOOC or wanted people outside of my field of view to be picking up my LOOCs. Usually loocs are done for silly little things that are relevent only in the here and now with those that can see the person.
> 
> SubtleLOOC feels like bloat to fix the symptom that your LOOC goes through walls, and LOOC going thru walls is just bad imo, because it can drag people into your looc wiggling that may not even be able to see you. I firmly believe it would be better to not have a SubtleLOOC, or an LOOC that goes thru walls at all. Dial it back to JUST traditional LOOC.
> 
> Less thought on needing to tab thru chat functions (which is a whole different basket of eggs, I think THAT thing needs a pop up to toggle what channels you can even tab thru to begin with, but that's another conversation entirely) more on raw simplicity. LOOC is used here in exactly the same way it is in 13, no reason for it to handle in some weird complex multi-channel way.

Okay so, subtle OOC is primarily intended to be used to ask for consent/clarification for extreme kinks, and I support keeping it. Even if LOOC is blocked by line of sight, I think a subtleOOC is still going to be useful. Probably. It certainly wouldn't be as relevant, though.

Superlagg committed:
removed localerOOC
made the los thing use occluision instead of obstruction!
moved the chatterthings to a floof folder

Tk420634 commented:
> > > I'm somewhat against adding an additional OOC channel - it's increasingly painful to tab through the various channels (my preferred way to switch my default channel) and I'm not sure the use case vs LOOC/subtleOOC. If anything, maybe LOOC could just get blocked by walls?
> > > Also, any unique files added should probably go in Floof or Floofstation folders (we haven't been super consistent as a community).
> > > This is super neat! Does stuff get blocked by windows too or just walls and opaque doors?
> > 
> > 
> > I brought this up some to Dan last night. SS13 has had LOOC for some time, at this point likely the better part of 8 or 9 years. It is blocked by walls, line of sight and all that. I have never, in all my time, felt like I needed a SubtleLOOC or wanted people outside of my field of view to be picking up my LOOCs. Usually loocs are done for silly little things that are relevent only in the here and now with those that can see the person.
> > SubtleLOOC feels like bloat to fix the symptom that your LOOC goes through walls, and LOOC going thru walls is just bad imo, because it can drag people into your looc wiggling that may not even be able to see you. I firmly believe it would be better to not have a SubtleLOOC, or an LOOC that goes thru walls at all. Dial it back to JUST traditional LOOC.
> > Less thought on needing to tab thru chat functions (which is a whole different basket of eggs, I think THAT thing needs a pop up to toggle what channels you can even tab thru to begin with, but that's another conversation entirely) more on raw simplicity. LOOC is used here in exactly the same way it is in 13, no reason for it to handle in some weird complex multi-channel way.
> 
> Okay so, subtle OOC is primarily intended to be used to ask for consent/clarification for extreme kinks, and I support keeping it. Even if LOOC is blocked by line of sight, I think a subtleOOC is still going to be useful. Probably. It certainly wouldn't be as relevant, though.

It may be, but tbh that was the same thing LOOC was used for.  That's why it sorta confuses me.  I'm hardly against SubtleLOOC existing entirely, it just feels odd to me because its main use case only seems to exist because LOOC sloppily goes through walls and informs others (outside of your scene space) that you want to clarify in what context you want your butt touched/untouched lmao

Superlagg committed:
i know the parentheses are unneeded
it just makes more sense to me this way

Memeji commented:
I mean, tbh I'm for making LOOC not go through walls if that can work. That way it works like the LocalerOOC but we don't add another OOC channel.

Think that'd work?

Superlagg commented:
Having an option to send LOOC through walls has been somewhat useful on Skyrat. It was used to ask people in scenes behind doors if they can come in, things like that. PDA messaging would work for that too, though we don't have that just yet.

I removed LocalerOOC since subtleOOC does that already, and left LOOC to go through walls. I can make it obey line of sight if needed tho.

Memeji commented:
> Having an option to send LOOC through walls has been somewhat useful on Skyrat. It was used to ask people in scenes behind doors if they can come in, things like that. PDA messaging would work for that too, though we don't have that just yet.
> 
> I removed LocalerOOC since subtleOOC does that already, and left LOOC to go through walls. I can make it obey line of sight if needed tho.

Was mainly throwing out an idea! If SubtleOOC works as LocalerOOC then we can leave LOOC to go through walls.

ShadesMars commented on Resources/Locale/en-US/Floof/random-barks/debug_chatterboy.ftl:
What did i just read...

MY GOD....
A!

Tk420634 replied:
I just looked at em, like.  Almost all of these are posts from people in the CB admin random talk zone lmfao, like the first loufa message was ME talking to someone else like two days ago

ShadesMars commented on Resources/Prototypes/Floof/Entities/Mobs/NPCs/debug.yml:
Intresting debug ent, as well, could be just a urist but this works.

Superlagg commented:
yay! (darn I dont get to see people ERP thru walls anymore)
